Compile issue on Arch Linux - libarchive

- Akira Version: master @ 1f976d41eeae7b310c66333ad8fbc029ee1d2206
- OS Info: Arch Linux rolling release

I'm trying to compile on Arch Linux, meson finds all the dependencies but it seems I have a different version of the Archive Vala package/library (sorry, I know almost nothing about Vala...) since the function signatures are different. What version of libarchive are you using? The documentation on [Valadoc](https://valadoc.org/libarchive/Archive.html) show indeed that the code is using the right parameters for the function calls, but that doesn't seem to match with what I have on my system.

pacman says that libarchive is at 3.3.3-1

Anyone can help me figure it out? Thanks!

```console
../src/FileFormat/ZipArchiveHandler.vala:268.20-268.86: error: 1 extra arguments for `Archive.Result Archive.Read.read_data_block (out weak uint8[], out Archive.int64_t)'
            while (archive.read_data_block (out buffer, out buffer_length, out offset) == Archive.Result.OK) {
                   ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
../src/FileFormat/ZipArchiveHandler.vala:269.49-269.54: error: Argument 1: Cannot convert from `void*' to `unowned uint8[]'
                if (extractor.write_data_block (buffer, buffer_length, offset) != Archive.Result.OK) {
                                                ^^^^^^
../src/FileFormat/ZipArchiveHandler.vala:269.21-269.78: error: 1 extra arguments for `ssize_t Archive.Write.write_data_block (uint8[], Archive.int64_t)'
                if (extractor.write_data_block (buffer, buffer_length, offset) != Archive.Result.OK) {
                    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
../src/FileFormat/ZipArchiveHandler.vala:319.37-319.57: error: Argument 1: Cannot convert from `int64' to `Archive.int64_t'
                    entry.set_size (file_info.get_size ());
                                    ^^^^^^^^^^^^^^^^^^^^^
../src/FileFormat/ZipArchiveHandler.vala:321.37-321.40: error: Argument 1: Cannot convert from `int' to `Archive.FileType'
                    entry.set_perm (0644);
                                    ^^^^
../src/FileFormat/ZipArchiveHandler.vala:336.25-336.63: error: 1 extra arguments for `ssize_t Archive.Write.write_data (uint8[])'
                        archive.write_data (buffer, bytes_read);
                        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
Compilation failed: 6 error(s), 0 warning(s)
ninja: build stopped: subcommand failed.
```

--- a/src/FileFormat/JsonObject.vala
+++ b/src/FileFormat/JsonObject.vala
@@ -31,7 +31,7 @@ public abstract class Akira.FileFormat.JsonObject : GLib.Object {
 
     private ObjectClass obj_class;
 
-    public JsonObject.from_object (Json.Object object) {
+    protected JsonObject.from_object (Json.Object object) {
         Object (object: object);
     }
 

--- a/src/FileFormat/JsonObjectArray.vala
+++ b/src/FileFormat/JsonObjectArray.vala
@@ -31,7 +31,7 @@ public abstract class Akira.FileFormat.JsonObjectArray : Object {
      *
      * Your JsonObject implementation should have it's own list of items
      */
-    public JsonObjectArray (Json.Object object, string property_name) {
+    protected JsonObjectArray (Json.Object object, string property_name) {
         Object (object: object, property_name: property_name);
     }
 

--- a/src/FileFormat/ZipArchiveHandler.vala
+++ b/src/FileFormat/ZipArchiveHandler.vala
@@ -266,11 +266,17 @@ public class Akira.FileFormat.ZipArchiveHandler : GLib.Object {
                 continue;
             }
 
+            Posix.off_t offset;
+#if VALA_0_42
+            uint8[] buffer;
+            while (archive.read_data_block (out buffer, out offset) == Archive.Result.OK) {
+                if (extractor.write_data_block (buffer, offset) != Archive.Result.OK) {
+#else
             void* buffer = null;
             size_t buffer_length;
-            Posix.off_t offset;
             while (archive.read_data_block (out buffer, out buffer_length, out offset) == Archive.Result.OK) {
                 if (extractor.write_data_block (buffer, buffer_length, offset) != Archive.Result.OK) {
+#endif
                     break;
                 }
             }
@@ -321,9 +327,15 @@ public class Akira.FileFormat.ZipArchiveHandler : GLib.Object {
                     // Add an entry to the archive
                     Archive.Entry entry = new Archive.Entry ();
                     entry.set_pathname (initial_folder.get_relative_path (current_file));
+#if VALA_0_42
+                    entry.set_size ((Archive.int64_t) file_info.get_size ());
+                    entry.set_filetype (Archive.FileType.IFREG);
+                    entry.set_perm (Archive.FileType.IFREG);
+#else
                     entry.set_size (file_info.get_size ());
                     entry.set_filetype ((uint) Posix.S_IFREG);
                     entry.set_perm (0644);
+#endif
 
                     if (archive.write_header (entry) != Archive.Result.OK) {
                         critical (
@@ -343,7 +355,11 @@ public class Akira.FileFormat.ZipArchiveHandler : GLib.Object {
                             break;
                         }
 
+#if VALA_0_42
+                        archive.write_data (buffer[0:bytes_read]);
+#else
                         archive.write_data (buffer, bytes_read);
+#endif
                     }
                 }
             }

